source/credentials_provider_profile.c

```diff
@@ -359,40 +372,44 @@ struct aws_credentials_provider *aws_credentials_provider_new_profile(
-    if (!credentials_file_path) {
-        AWS_LOGF_ERROR(
-            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
-            "static: Profile credentials parser failed resolve credentials file path");
-        goto on_finished;
-    }
-
-    config_file_path = aws_get_config_file_path(allocator, &options->config_file_name_override);
-    if (!config_file_path) {
-        AWS_LOGF_ERROR(
-            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "static: Profile credentials parser failed resolve config file path");
-        goto on_finished;
-    }
-
     profile_name = aws_get_profile_name(allocator, &options->profile_name_override);
     if (!profile_name) {
         AWS_LOGF_ERROR(
             AWS_LS_AUTH_CREDENTIALS_PROVIDER, "static: Profile credentials parser failed to resolve profile name");
         goto on_finished;
     }
-    config_profiles = aws_profile_collection_new_from_file(allocator, config_file_path, AWS_PST_CONFIG);
-    credentials_profiles = aws_profile_collection_new_from_file(allocator, credentials_file_path, AWS_PST_CREDENTIALS);
 
-    if (!(config_profiles || credentials_profiles)) {
-        AWS_LOGF_ERROR(
-            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
-            "static: Profile credentials parser could not load or parse"
-            " a credentials or config file.");
-        goto on_finished;
-    }
+    merged_profiles = aws_profile_collection_acquire(options->profile_collection_cached);
+    if (!merged_profiles) {
+        credentials_file_path = aws_get_credentials_file_path(allocator, &options->credentials_file_name_override);
+        if (!credentials_file_path) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "static: Profile credentials parser failed resolve credentials file path");
+            goto on_finished;
+        }
+
+        config_file_path = aws_get_config_file_path(allocator, &options->config_file_name_override);
+        if (!config_file_path) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER, "static: Profile credentials parser failed resolve config file path");
+            goto on_finished;
+        }
 
-    merged_profiles = aws_profile_collection_new_from_merge(allocator, config_profiles, credentials_profiles);
+        config_profiles = aws_profile_collection_new_from_file(allocator, config_file_path, AWS_PST_CONFIG);
+        credentials_profiles =
+            aws_profile_collection_new_from_file(allocator, credentials_file_path, AWS_PST_CREDENTIALS);
+
+        if (!(config_profiles || credentials_profiles)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "static: Profile credentials parser could not load or parse"
+                " a credentials or config file.");
+            goto on_finished;
+        }
 
+        merged_profiles = aws_profile_collection_new_from_merge(allocator, config_profiles, credentials_profiles);
```

Comment:
possible bug in both old and new code ....

if only 1/2 of the files is successfully loaded, we still call `aws_profile_collection_new_from_merge()`

not sure what the idea behavior is:
- is it OK to only load 1/2 files?
- if it's OK, should we silently ignore the error like we do now, even when a filepath override is passed in?
- is it OK to call merge when 1/2 profiles is NULL?

Reply:
I think it is okay to load only 1/2 of the files. We will attempt to load properties from the file and generate an error if any necessary property is missing. In `aws_profile_collection_new_from_merge`, we do handle 1/2 of the files.

```
    if (config_profiles != NULL) {
        if (s_profile_collection_merge(merged, config_profiles)) {
            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_PROFILE, "Failed to merge config profile set");
            goto cleanup;
        }
    }

    if (credentials_profiles != NULL) {
        if (s_profile_collection_merge(merged, credentials_profiles)) {
            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_PROFILE, "Failed to merge credentials profile set");
            goto cleanup;
        }
    }
```